crates/trie/sparse-parallel/src/trie.rs

```diff
@@ -841,6 +980,198 @@ impl SparseSubtrie {
+    /// collapsing extension nodes as needed.
+    ///
+    /// # Returns
+    ///
+    /// Returns the `Ok` if the update is successful.
+    /// If a split branch was added this is returned as well, along with its path.
+    ///
+    /// Note: If an update requires revealing a blinded node, an error is returned if the blinded
+    /// provider returns an error.
+    pub fn update_leaf(
+        &mut self,
+        path: Nibbles,
+        value: Vec<u8>,
+        provider: impl BlindedProvider,
+    ) -> SparseTrieResult<()> {
+        debug_assert!(path.starts_with(&self.path));
+        let existing = self.inner.values.insert(path, value);
+        if existing.is_some() {
+            // trie structure unchanged, return immediately
+            return Ok(())
+        }
+
+        // Here we are starting at the root of the subtrie, and traversing from there.
+        let mut current = Some(self.path);
+        while let Some(current_path) = current {
+            match self.next_node(current_path, &path, &provider)? {
+                LeafUpdateStep::Continue { next_node } => {
+                    current = Some(next_node);
+                }
+                LeafUpdateStep::Complete { .. } | LeafUpdateStep::NoOp => {
+                    current = None;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Processes the current node, returning what to do next in the leaf update process.
+    ///
+    /// This will add or update any nodes in the trie as necessary.
+    ///
+    /// Returns a `LeafUpdateStep` containing the next node to process (if any) and
+    /// the paths of nodes that were inserted during this step.
+    fn next_node(
+        &mut self,
+        mut current: Nibbles,
+        path: &Nibbles,
+        provider: impl BlindedProvider,
+    ) -> SparseTrieResult<LeafUpdateStep> {
+        debug_assert!(path.starts_with(&self.path));
+        debug_assert!(current.starts_with(&self.path));
+        debug_assert!(path.starts_with(&current));
+        let Some(node) = self.nodes.get_mut(&current) else {
+            return Ok(LeafUpdateStep::NoOp);
+        };
+        match node {
+            SparseNode::Empty => {
+                // We need to insert the node with a different path and key depending on the path of
+                // the subtrie.
+                let path = path.slice(self.path.len()..);
+                *node = SparseNode::new_leaf(path);
+                Ok(LeafUpdateStep::complete_with_insertions(vec![current]))
+            }
+            &mut SparseNode::Hash(hash) => {
+                Err(SparseTrieErrorKind::BlindedNode { path: current, hash }.into())
+            }
+            SparseNode::Leaf { key: current_key, .. } => {
+                current.extend(current_key);
+
+                // this leaf is being updated
+                if &current == path {
+                    unreachable!("we already checked leaf presence in the beginning");
+                }
```

Comment:
should this be a debug assertion?

Reply:
Yes probably, tbh I am not 100% sure when we would hit this

Reply:
changed to `debug_assert!`

Reply:
test